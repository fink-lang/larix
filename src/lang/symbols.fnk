{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
