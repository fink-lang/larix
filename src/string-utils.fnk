unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
