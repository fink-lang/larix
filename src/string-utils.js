import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

