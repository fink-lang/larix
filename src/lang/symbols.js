import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
