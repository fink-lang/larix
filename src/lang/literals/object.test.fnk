{describe, it, eq, to_throw, snapshot} = import '../../testing/jest'

{parse_expr} = import '../../'


describe:: 'object {...}', fn:

  it:: 'parses empty: {}', fn:
    parse_expr(`{}`) eq snapshot


  it:: 'parses single shorthand prop: {foo}', fn:
    parse_expr(`{foo}`) eq snapshot


  it:: 'parses multiple shorthand exprs: {foo, bar}', fn:
    parse_expr(`{foo, bar}`) eq snapshot


  it:: 'parses multiple shorthand exprs: {foo, bar}', fn:
    parse_expr(`{foo, bar: 123, shrub: 'ni'}`) eq snapshot


  it:: 'parses single prop: {foo: spam}', fn:
    parse_expr(`{foo: spam}`) eq snapshot


  it:: 'parses spread prop: {foo, ...bar}', fn:
    parse_expr(`{foo, ...bar}`) eq snapshot


  it:: `parses single str prop: {'foo': spam}`, fn:
    parse_expr(`{'foo': spam}`) eq snapshot


  it:: 'parses default assignment prop: {foo=123}', fn:
    parse_expr(`{foo=123}`) eq snapshot


  it:: `parses calculated prop: {(Symbol('foo')): 123, ...}`, fn:
    parse_expr(`{foo, (Symbol('bar')): 123, spam}`) eq snapshot


  it:: 'parses single prop with block value: {foo: spam ...}', fn:
    parse_expr(`
      {
        foo:
          spam
          ni
      }
    `) eq snapshot



describe:: 'object {...} - parsing failures', fn:
  it:: `throws when missing '}'`, fn:
    (
      fn: parse_expr(`{`, 'test.fnk')
    ) to_throw `
      test.fnk:1:1
      1| {
          ^

      Expected '}' but found Symbol(end).
    `

  it:: `throws when missing '}'`, fn:
    (
      fn: parse_expr(`{foo) bar}`, 'test.fnk')
    ) to_throw `
      test.fnk:1:4
      1| {foo) bar}
             ^

      Expected '}' but found ')'.
    `
