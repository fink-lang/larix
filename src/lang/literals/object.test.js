import {other_token} from '@fink/prattler/symbols';

import {parse_expr} from '../../';


import {strip_block} from '../../string-utils';


describe('object {...}', ()=> {

  it('parses empty: {}', ()=> {
    expect(
      parse_expr(`{}`)
    ).toEqual({
      type: 'object',
      exprs: [],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 2, line: 1, column: 2}
      }
    });
  });


  it('parses single shorthand prop: {foo}', ()=> {
    expect(
      parse_expr(`{foo}`)
    ).toEqual({
      type: 'object',
      exprs: [{
        type: 'prop',
        key: parse_expr(` foo`),
        value: parse_expr(` foo`),
        loc: {
          start: {pos: 1, line: 1, column: 1},
          end: {pos: 4, line: 1, column: 4}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 5, line: 1, column: 5}
      }
    });
  });


  it('parses multiple shorthand exprs: {foo, bar}', ()=> {
    expect(
      parse_expr(`{foo, bar}`)
    ).toEqual({
      type: 'object',
      exprs: [{
        type: 'prop',
        key: parse_expr(` foo`),
        value: parse_expr(` foo`),
        loc: {
          start: {pos: 1, line: 1, column: 1},
          end: {pos: 4, line: 1, column: 4}
        }
      }, {
        type: 'prop',
        key: parse_expr(`      bar`),
        value: parse_expr(`      bar`),
        loc: {
          start: {pos: 6, line: 1, column: 6},
          end: {pos: 9, line: 1, column: 9}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 10, line: 1, column: 10}
      }
    });
  });


  it('parses single prop: {foo: spam}', ()=> {
    expect(
      parse_expr(`{foo: spam}`)
    ).toEqual({
      type: 'object',
      exprs: [{
        type: 'prop',
        key: parse_expr(` foo`),
        value: parse_expr(`    : spam`),
        loc: {
          start: {pos: 1, line: 1, column: 1},
          end: {pos: 10, line: 1, column: 10}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 11, line: 1, column: 11}
      }
    });
  });


  it('parses single str prop: {`foo`: spam}', ()=> {
    expect(
      parse_expr('{`foo`: spam}')
    ).toEqual({
      type: 'object',
      exprs: [{
        type: 'prop',
        key: parse_expr(' `foo`'),
        value: parse_expr(`      : spam`),
        loc: {
          start: {pos: 1, line: 1, column: 1},
          end: {pos: 12, line: 1, column: 12}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 13, line: 1, column: 13}
      }
    });
  });


  it('parses default assignment prop: {foo=123}', ()=> {
    expect(
      parse_expr(`{foo=123}`)
    ).toEqual({
      type: 'object',
      exprs: [{
        type: 'prop',
        key: parse_expr(` foo`),
        value: parse_expr(` foo=123`),
        loc: {
          start: {pos: 1, line: 1, column: 1},
          end: {pos: 8, line: 1, column: 8}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 9, line: 1, column: 9}
      }
    });
  });
});


describe('object {...} - parsing failures', ()=> {
  it('throws when missing `}`', ()=> {
    expect(
      ()=> parse_expr(`{`)
    ).toThrow(strip_block`
      Expected '}' but found Symbol(end):
      1| {
          ^`
    );
  });

  // TODO: used to fail
  // it('throws when missing value after`:`', ()=> {
  //   expect(
  //     ()=> parse_expr(`{foo:,}`)
  //   ).toThrow(strip_block`
  //     Cannot use ',' as start of expression:
  //     1| {foo:,}
  //             ^`
  //   );
  // });

  it('throws when missing `,`', ()=> {
    expect(
      ()=> parse_expr(`{foo) bar}`)
    ).toThrow(strip_block`
      Expected ',' but found ')':
      1| {foo) bar}
             ^`
    );
  });
});

