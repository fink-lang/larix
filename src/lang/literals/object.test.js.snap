// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object {...} parses default assignment prop: {foo=123} 1`] = `
object (1:0-1:9)
  prop (1:1-1:8)
    other (1:1-1:4) foo
    assign = (1:1-1:8)
      other (1:1-1:4) foo
      other (1:5-1:8) 123
`;

exports[`object {...} parses empty: {} 1`] = `
object (1:0-1:2)

`;

exports[`object {...} parses multiple shorthand exprs: {foo, bar} 1`] = `
object (1:0-1:10)
  prop (1:1-1:4)
    other (1:1-1:4) foo
    other (1:1-1:4) foo
  prop (1:6-1:9)
    other (1:6-1:9) bar
    other (1:6-1:9) bar
`;

exports[`object {...} parses multiple shorthand props: {foo, bar} 1`] = `
object (1:0-1:10)
  prop (1:1-1:4)
    other (1:1-1:4) foo
    other (1:1-1:4) foo
  prop (1:6-1:9)
    other (1:6-1:9) bar
    other (1:6-1:9) bar
`;

exports[`object {...} parses single prop: {foo: spam} 1`] = `
object (1:0-1:11)
  prop (1:1-1:10)
    other (1:1-1:4) foo
    block (1:4-1:10)
      other (1:6-1:10) spam
`;

exports[`object {...} parses single shorthand prop: {foo} 1`] = `
object (1:0-1:5)
  prop (1:1-1:4)
    other (1:1-1:4) foo
    other (1:1-1:4) foo
`;

exports[`object {...} parses single str prop: {\`foo\`: spam} 1`] = `
object (1:0-1:13)
  prop (1:1-1:12)
    string \` (1:1-1:6)
      \`foo\`
    block (1:6-1:12)
      other (1:8-1:12) spam
`;
